lib/Geo/ShapeContainer.cpp

```diff
@@ -668,4 +675,131 @@ std::vector<S2CellId> ShapeContainer::covering(S2RegionCoverer& coverer) const {
+  // because smaller S2Shape -- point needs 30 but really only 25 so it's enough
+  encoder.Ensure(30);
+  encoder.put8(serialize(_type));
+  switch (_type) {
+    case Type::S2_POINT: {
+      auto const& data = basics::downCast<S2PointRegion>(*_data);
+      encodePoint(encoder, data.point(), hint);
+    } break;
+    case Type::S2_POLYLINE: {
+      auto const& data = basics::downCast<S2Polyline>(*_data);
+      if (hint == s2coding::CodingHint::FAST) {
+        data.EncodeUncompressed(&encoder);
+      } else {
+        data.EncodeMostCompact(&encoder);
+      }
+    } break;
+    case Type::S2_POLYGON: {
+      auto const& data = basics::downCast<S2Polygon>(*_data);
+      if (hint == s2coding::CodingHint::FAST) {
+        data.EncodeUncompressed(&encoder);
+      } else {
+        data.Encode(&encoder);
+      }
+    } break;
+    case Type::S2_MULTIPOINT: {
+      auto const& data = basics::downCast<S2MultiPointRegion>(*_data);
+      data.Encode(&encoder, hint);
+    } break;
+    case Type::S2_MULTIPOLYLINE: {
+      auto const& data = basics::downCast<S2MultiPolyline>(*_data);
+      data.Encode(&encoder, hint);
+    } break;
+    case Type::S2_LATLNGRECT:  // not exist in non-legacy geo
+    case Type::EMPTY:
+      TRI_ASSERT(false);
+  }
+}
+
+template<ShapeContainer::Type Type, typename T>
+bool ShapeContainer::DecodeImpl(Decoder& decoder) {
+  if (ADB_UNLIKELY(_type != Type)) {
+    if constexpr (std::is_same_v<T, S2PointRegion>) {
+      _data = std::make_unique<T>(S2Point{});
+    } else {
+      _data = std::make_unique<T>();
+    }
+    _type = Type;
+  }
+  if constexpr (std::is_same_v<T, S2PointRegion>) {
+    return decodePoint(decoder, basics::downCast<T>(*_data));
+  } else {
+    return basics::downCast<T>(*_data).Decode(&decoder);
+  }
+}
+
+bool ShapeContainer::Decode(Decoder& decoder) {
+  if (decoder.avail() < sizeof(uint8_t)) {
+    return false;
+  }
+  auto type = decoder.get8();
+  switch (type) {
+    case serialize(Type::S2_POINT):
+      return DecodeImpl<Type::S2_POINT, S2PointRegion>(decoder);
+    case serialize(Type::S2_POLYLINE):
+      return DecodeImpl<Type::S2_POLYLINE, S2Polyline>(decoder);
+    case serialize(Type::S2_POLYGON):
+      return DecodeImpl<Type::S2_POLYGON, S2Polygon>(decoder);
+    case serialize(Type::S2_MULTIPOINT):
+      return DecodeImpl<Type::S2_MULTIPOINT, S2MultiPointRegion>(decoder);
+    case serialize(Type::S2_MULTIPOLYLINE):
+      return DecodeImpl<Type::S2_MULTIPOLYLINE, S2MultiPolyline>(decoder);
+    default:
+      TRI_ASSERT(false);
+  }
+  return false;
+}
+
+void encodePoint(Encoder& encoder, S2Point const& point,
+                 s2coding::CodingHint hint) {
+  TRI_ASSERT(S2::IsUnitLength(point));
+  TRI_ASSERT(encoder.avail() >= sizeof(uint8_t) + 3 * sizeof(double));
+  // if (hint == s2coding::CodingHint::FAST) {
+  encoder.put8(kCurrentLosslessEncodingVersionNumber);
+  for (int i = 0; i < 3; ++i) {
+    encoder.putdouble(point[i]);
+  }
+  /* } else {
```

Comment:
Leftovers? Otherwise please add a TODO when this should be uncommented

Reply:
We not really sure about it yes